src/lib.rs

```diff
@@ -80,9 +80,11 @@ impl MmioUart {
     pub fn pointer_to_data(&mut self) -> *mut u32 {
         unsafe { &raw mut (*self.ptr).data }
     }
-    pub fn read_data(&mut self) -> u32 {
-        let addr = self.pointer_to_data();
-        unsafe { addr.read_volatile() }
+    pub fn read_data(&self) -> u32 {
+        let addr = unsafe { core::ptr::addr_of!((*self.ptr).data) };
```

Comment:
```suggestion
        let addr = self.pointer_to_data();
```

Sorry, there's always one more thing.

Reply:
The mix of &raw mut in the function above and addr_of! here just seemed incongruous.

Reply:
hmm, I did this because the actual code now looks like this for a PureRead register. I can not use pointer_to_data because it is mutable.

Reply:
Ah. Hmm. Do we need to require exclusive access to get a pointer? Probably not?

Reply:
No.. it probably is there to enforce ownership rules.. and it is a public function after all. One option would be to have a mutable and non-mutable pointer getter method where one returns `*const` and the other returns `*mut` .

Reply:
For example:

`pointer_mut_to_#ident` and `pointer_to_#ident`.

Reply:
You can have a look at  https://github.com/us-irs/derive-mmio/commit/08ad1bc57b0f88a3d60f3e2efb572c2c875e219a where I tried this.

Reply:
> But a safe non-mutable getter would now return a mutable pointer, which seems wrong.

It seems wrong, but pointers are unsafe, and you can unsafely create yourself a second handle and use *that* to call any `&mut self` method you like, so I think I'm OK with returning pointers from a &self handle.

But in any case, we can do that in a future PR. I understand the reason for the duplicate code, and it's fine.

Reply:
Ah that's right, doing anything with the pointer requires unsafe anyway..

Reply:
And thinking about it, that's exactly how `core::cell::UnsafeCell` works.